### PR TITLE
Add assessment highlighting to Apollon

### PR DESCRIPTION
**Pull request by nicolasruscher**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Currently Apollon does not support highlighting the feedback within an exams second correction round

### Description

We add the possibility to highlight feedback correction rounds into the modeling assessment editor and Apollon. 
Please note: online mergeable / testable after Artemis uses Apollon > 2.7.0 

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration -> Exam with second correction round -> modeling exercise second correction
3. You should see the Difference View Button. Clicking on it shows a new badge for every feedback type (Apollon + unreferenced Feedback). The Badge states which correction round the feedback was taken in. 
4. Please Note: Apollon Feedback Badges are only refreshed with new information once the button is switched off and on again due to Apollon Artemis communication

### Screenshots

Bildschirmfoto 2021-06-01 um 07 57 08
Bildschirmfoto 2021-06-01 um 08 03 43

### CODE REVIEW

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts [hunk 1]
```diff
@@ -270,9 +281,27 @@ export class ModelingAssessmentComponent implements AfterViewInit, OnDestroy, On
...
         if (this.apollonEditor) {
             this.apollonEditor!.model = this.model;
         }
     }
+
+    private calculateLabel(feedback: any) {
+        if (this.highlightDifferences && this.languageService.getCurrentLanguage() === 'en') {
+            return feedback.copiedFeedbackId ? 'First correction round' : 'Second correction round';
+        } else if (this.highlightDifferences && this.languageService.getCurrentLanguage() === 'de') {
+            return feedback.copiedFeedbackId ? 'Erste Korrekturrunde' : 'Zweite Korrekturrunde';
```
Comment (Malyuk-A) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:296]: This seems to work, but also like a hardcoded version of the Artemis translation pipe that we should use instead? 
(I am not sure, that is why I am asking)
image
If this is ok, when (in which cases) should we use the Artemis translation pipe instead?

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts [hunk 2]
```diff
@@ -270,9 +281,27 @@ export class ModelingAssessmentComponent implements AfterViewInit, OnDestroy, On
...
+        }
+        return undefined;
+    }
+
+    private calculateLabelColor(feedback: any) {
+        if (this.highlightDifferences) {
+            return feedback.copiedFeedbackId ? '#3e8acc' : '#ffa561';
```
Comment (Malyuk-A) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:303]: The two colors seem to be instances of hardcoded magic numbers. 
Wouldn't it be better to define two constants like `const firstCorrectionRoundColor = #3e8acc`, etc., or even create new (global) `enums` so you could reuse them in your test class as well?

Or is it common in Artemis to use colors like magic numbers?

File: src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts [hunk 3]
```diff
@@ -215,6 +221,43 @@ describe('ModelingAssessmentComponent', () => {
...
+        fixture.detectChanges();
+        comp.feedbacks = [mockFeedbackWithReference];
+        comp.referencedFeedbacks = [mockFeedbackWithReference];
+
+        comp.ngOnChanges(changes);
+
+        expect(comp.apollonEditor).to.exist;
+        const apollonModel = comp.apollonEditor!.model;
+        const assessments: any = apollonModel.assessments;
+        expect(assessments[0].labelColor).to.equal('#ffa561');
```
Comment (Malyuk-A) [inline src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts:237]: See my comment about color and magic numbers above.

File: src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts [hunk 4]
```diff
@@ -215,6 +221,43 @@ describe('ModelingAssessmentComponent', () => {
...
+        fixture.detectChanges();
+        comp.feedbacks = [mockFeedbackWithReferenceCopied];
+        comp.referencedFeedbacks = [mockFeedbackWithReferenceCopied];
+
+        comp.ngOnChanges(changes);
+
+        expect(comp.apollonEditor).to.exist;
+        const apollonModel = comp.apollonEditor!.model;
+        const assessments: any = apollonModel.assessments;
+        expect(assessments[0].labelColor).to.equal('#3e8acc');
```
Comment (Malyuk-A) [inline src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts:256]: Same here.

**Review by Malyuk-A — CHANGES_REQUESTED** [review]
Tested on Ts1. Works as expected.
I have a couple of questions regarding the code. (See my code comments).

File: src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts (hunk 4 above)
Comment (nicolasruscher) [reply src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts]: Done

File: src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts (hunk 3 above)
Comment (nicolasruscher) [reply src/test/javascript/spec/component/modeling-assessment/modeling-assessment.component.spec.ts]: Done

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts (hunk 2 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts]: Good finding, I refactored this and moved all definitions into class variables

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts (hunk 1 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts]: After some testing I got the pipe to work on typescript files as well, its being used similar to the inline html now, thanks!

**Review by Malyuk-A — APPROVED** [review]
Thank you for refactoring the code, looks nice.👍

**Review by ivohashamov — APPROVED** [review]
Code changes make sense!

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts [hunk 5]
```diff
@@ -270,9 +284,27 @@ export class ModelingAssessmentComponent implements AfterViewInit, OnDestroy, On
...
             feedback: feedback.text || undefined,
+            label: this.calculateLabel(feedback),
+            labelColor: this.calculateLabelColor(feedback),
         }));
         if (this.apollonEditor) {
             this.apollonEditor!.model = this.model;
         }
     }
+
+    private calculateLabel(feedback: any) {
```
Comment (imstreetspirit) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:295]: can we avoid to use `any` in here? i'm just asking :)

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts [hunk 6]
```diff
@@ -270,9 +284,27 @@ export class ModelingAssessmentComponent implements AfterViewInit, OnDestroy, On
...
+        const firstCorrectionRoundText = this.artemisTranslatePipe.transform('artemisApp.assessment.diffView.correctionRoundDiffFirst');
+        const secondCorrectionRoundText = this.artemisTranslatePipe.transform('artemisApp.assessment.diffView.correctionRoundDiffSecond');
+        if (this.highlightDifferences) {
+            return feedback.copiedFeedbackId ? firstCorrectionRoundText : secondCorrectionRoundText;
+        }
+        return undefined;
+    }
+
+    private calculateLabelColor(feedback: any) {
```
Comment (imstreetspirit) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:304]: same as above...

**Review by ArtemTsurikov — DISMISSED** [review]
Tested on ts1, looks good

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts (hunk 5 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:295]: sadly I have not found a way, as the datatype comes from Apollon and Artemis can't infer what type it is.

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts (hunk 6 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:304]: sadly I have not found a way, as the datatype comes from Apollon and Artemis can't infer what type it is

**Review by maximilianbergmann — APPROVED** [review]
Tested on TS3. Works as expected

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts [hunk 7]
```diff
@@ -4,7 +4,9 @@ import { JhiAlertService } from 'ng-jhipster';
 import interact from 'interactjs';
 import { Feedback, FeedbackType } from 'app/entities/feedback.model';
 import * as $ from 'jquery';
+import { JhiLanguageService } from 'ng-jhipster';
```
Comment (niclasschuemann) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:7]: According to GitHub Actions / client-style, 'JhiLanguageService' is defined but never used. Should this be fixed?

**Review by niclasschuemann — COMMENTED** [review]
The code looks good to me, besides the one comment above I'm not sure about.

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts (hunk 7 above)
Comment (nicolasruscher) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:7]:
```suggestion
```

Comment (ndricimrr) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment.component.ts:7]:
```suggestion
```

**Review by MDK5 — DISMISSED** [review]
Tested in testing session on ts1, worked as described

**Review by PatZintel — APPROVED** [review]
worked on ts1

**Review by ndricimrr — APPROVED** [review]
Tested during testing session on TS1. Works as expected

**Review by niclasschuemann — APPROVED** [review]
Re-approve

**Review by FlorianGlombik — APPROVED** [review]
Tested it during the test session on ts1, the badges are added as described

**Review by canyelten — APPROVED** [review]
Tested on TS2, works as described.

**Review by lschlesinger — APPROVED** [review]
I tested it during Testing Session! Works as intended!

**Review by argertboja — APPROVED** [review]
Tested on TS2, works as expected!

**Review by jpbernius — APPROVED** [review]

**Pull request merged**